### ISSUE TITLE
Investigate "JavaScript heap out of memory" error

### ISSUE THREAD

**Issue by bqbn** [mozilla/addons-frontend#3916]
### Describe the problem and steps to reproduce it:
We had increased amount of traffic this morning, and the node crashed with "JavaScript heap out of memory" error. When it happened, the docker daemon automatically restarted the container, however, shortly after the restart the node crashed again.

On one of the instances, we see a series of crash like this:

```
$ sudo journalctl -u docker-addons-frontend.service |egrep "JavaScript heap out of memory"
Nov 15 14:11:42 ip-172-31-28-11 docker[2974]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:28:42 ip-172-31-28-11 docker[23138]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:38:08 ip-172-31-28-11 docker[24671]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:45:17 ip-172-31-28-11 docker[26213]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:54:37 ip-172-31-28-11 docker[27466]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 15:10:45 ip-172-31-28-11 docker[29064]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
```

In order to reduce the # of 500 caused by the issue, we had to bump up the # of instances in the cluster from 8 to 40. And that eventually stopped the "heap oom" from continuously happening.
 

### What happened?
As above.

### What did you expect to happen?
The purpose for this ticket is to
1. investigate what may have caused the "heap oom". For example, could it be memory leak somewhere?
2. possible solutions going forward. For example increasing heap size.

### Anything else we should know?
As above, but please let us know if more info is needed.

**Comment by bqbn** [comment]
To add some more info, we had 8 instances in the cluster before the incident. And here is the timing of when "heap oom" happened on them,

```
Nov 15 14:06:06 ip-172-31-11-84 docker[2954]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:06:41 ip-172-31-22-109 docker[2958]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:07:04 ip-172-31-15-172 docker[2963]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:07:11 ip-172-31-34-50 docker[2957]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:10:18 ip-172-31-23-97 docker[2973]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:11:42 ip-172-31-28-11 docker[2974]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:13:14 ip-172-31-22-109 docker[31049]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:13:22 ip-172-31-11-84 docker[31019]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:19:47 ip-172-31-23-97 docker[31658]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:21:19 ip-172-31-11-84 docker[32265]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:21:53 ip-172-31-22-109 docker[32213]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:23:07 ip-172-31-34-50 docker[31331]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:27:47 ip-172-31-5-114 docker[2964]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:28:29 ip-172-31-23-97 docker[686]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:28:42 ip-172-31-28-11 docker[23138]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:33:14 ip-172-31-11-84 docker[1162]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:37:06 ip-172-31-23-97 docker[2113]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:38:08 ip-172-31-28-11 docker[24671]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:40:25 ip-172-31-44-111 docker[2955]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:42:10 ip-172-31-11-84 docker[3005]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:45:17 ip-172-31-28-11 docker[26213]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:45:26 ip-172-31-23-97 docker[3531]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:48:47 ip-172-31-44-111 docker[3938]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:49:35 ip-172-31-11-84 docker[4461]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:54:37 ip-172-31-28-11 docker[27466]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:56:07 ip-172-31-44-111 docker[5314]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:59:15 ip-172-31-34-50 docker[1153]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 15:04:19 ip-172-31-44-111 docker[6516]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 15:09:09 ip-172-31-22-109 docker[1185]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 15:10:45 ip-172-31-28-11 docker[29064]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
```

All 8 had "heap oom", and below is the # of times it happened to them:

```
   7 ip-172-31-28-11
   6 ip-172-31-11-84
   5 ip-172-31-23-97
   4 ip-172-31-44-111
   4 ip-172-31-22-109
   3 ip-172-31-34-50
   1 ip-172-31-5-114
   1 ip-172-31-15-172
```

**Comment by kumar303** [comment]
<details>
<summary>
Here's an example of a crash
</summary>

```
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: <--- Last few GCs --->
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 372466 ms: Mark-sweep 1254.1 (1435.8) -> 1253.6 (1435.8) MB, 1703.9 / 0.1 ms [allocation failure] [GC in old space requested].
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 374150 ms: Mark-sweep 1253.6 (1435.8) -> 1253.6 (1435.8) MB, 1683.4 / 0.2 ms [allocation failure] [GC in old space requested].
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 375994 ms: Mark-sweep 1253.6 (1435.8) -> 1257.4 (1406.8) MB, 1844.1 / 0.2 ms [last resort gc].
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 377751 ms: Mark-sweep 1257.4 (1406.8) -> 1262.6 (1406.8) MB, 1756.2 / 0.2 ms [last resort gc].
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: <--- JS stacktrace --->
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: ==== JS stack trace =========================================
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: Security context: 0x396f50dcf781 <JS Object>
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 1: addTask(aka addTask) [/srv/node/node_modules/redux-saga/lib/internal/proc.js:~93] [pc=0x1081174e12ba] (this=0x396f50d04381 <undefined>,task=0x1b1cecbb83b1 <an Object with map 0xd540bfb8be9>)
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 2: forkQueue(aka forkQueue) [/srv/node/node_modules/redux-saga/lib/internal/proc.js:~82] [pc=0x10812a857d62] (this=0x396f50d04381 <undefined>,name=0x1246daf4bdb1 <String[9]: abuseSaga>,mainTask=0...
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 1: node::Abort() [node]
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 2: 0x10a04bc [node]
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 3: v8::Utils::ReportApiFailure(char const*, char const*) [node]
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [node]
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 5: v8::internal::Handle<v8::internal::JSFunction> v8::internal::Factory::New<v8::internal::JSFunction>(v8::internal::Handle<v8::internal::Map>, v8::internal::AllocationSpace) [node]
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 6: v8::internal::Factory::NewFunction(v8::internal::Handle<v8::internal::Map>, v8::internal::Handle<v8::internal::SharedFunctionInfo>, v8::internal::Handle<v8::internal::Context>, v8::internal::PretenureFlag) [node]
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 7: v8::internal::Factory::NewFunctionFromSharedFunctionInfo(v8::internal::Handle<v8::internal::SharedFunctionInfo>, v8::internal::Handle<v8::internal::Context>, v8::internal::PretenureFlag) [node]
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 8: v8::internal::Runtime_NewClosure_Tenured(int, v8::internal::Object**, v8::internal::Isolate*) [node]
Nov 15 14:19:39 ip-172-31-28-11 docker[21727]: 9: 0x108115a092a7
Nov 15 14:20:07 ip-172-31-28-11 docker[21727]: Aborted (core dumped)
Nov 15 14:20:07 ip-172-31-28-11 docker[21727]: error Command failed with exit code 134.
Nov 15 14:20:07 ip-172-31-28-11 docker[21727]: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Nov 15 14:20:09 ip-172-31-28-11 systemd[1]: docker-addons-frontend.service: main process exited, code=exited, status=1/FAILURE
Nov 15 14:20:09 ip-172-31-28-11 docker[23108]: addons-frontend
Nov 15 14:20:09 ip-172-31-28-11 docker[23115]: addons-frontend
Nov 15 14:20:09 ip-172-31-28-11 systemd[1]: Unit docker-addons-frontend.service entered failed state.
Nov 15 14:20:09 ip-172-31-28-11 systemd[1]: docker-addons-frontend.service failed.
Nov 15 14:20:09 ip-172-31-28-11 systemd[1]: docker-addons-frontend.service holdoff time over, scheduling restart.
Nov 15 14:20:09 ip-172-31-28-11 systemd[1]: Starting Daemon for addons_frontend...
Nov 15 14:20:09 ip-172-31-28-11 docker[23124]: Error response from daemon: Cannot kill container addons-frontend: No such container: addons-frontend
Nov 15 14:20:09 ip-172-31-28-11 docker[23131]: Error response from daemon: No such container: addons-frontend
Nov 15 14:20:09 ip-172-31-28-11 systemd[1]: Started Daemon for addons_frontend.
```

</details>

**Comment by tofumatt** [comment]
Time to rewrite it in Rust.

**Comment by kumar303** [comment]
Our NodeJS instances were taking up to 5 minutes to fully process requests at the time of the OOM crashes. Nginx would have responded to the client within 30 seconds with a 503 but the app would finish processing the response in the background.

<img width="1277" alt="screenshot 2017-11-15 15 42 23" src="https://user-images.githubusercontent.com/55398/32862155-cffd1b42-ca1c-11e7-89b0-1bd60f353580.png">

**Comment by kumar303** [comment]
**tl;dr** The following analysis makes me think we don't have a memory leak; we just had very high load.

I got a log file from jason that had 7 OOM crashes in it. When I analyzed the window of activity between each one, they were serving 2800 - 3300 concurrent requests right before crashing. This sounds like a lot (the app normally does 200 requests/sec) but the ELB graph up above suggests some responses were taking 5 minutes to process so a large backlog seems possible. 

Each response gets a `store` instance which lives for the duration of a response. The `store` contains the entire application state and gets quite large so that could be the culprit (just a theory).

I used this script if anyone wants to take a look to make sure I was checking [`server/base.js`](https://github.com/mozilla/addons-frontend/blob/e298b61185bbe4f8e319bae47777423f3ee17218/src/core/server/base.js) messages correctly. https://gist.github.com/kumar303/bb0ef2ab9f3865e2f2f7edd6b7539c59

Example output between the first and second crash:

```
{ errors: 0,
  firstRender: 4286,
  heapOOM: 7,
  highestConcurrentRequests: 3344,
  finalConcurrentRequests: 3344,
  secondRender: 942,
  total: 39370 }
```

**Comment by kumar303** [comment]
I'm pretty certain that we don't have a memory leak, which was my main concern. I think I investigated this enough to close it but we may learn more things in the future. The problem with analyzing these logs is the server seemed way over capacity. We need to find out what its capacity is. Here are the follow-ups:

- Determine load capacity: https://github.com/mozilla/addons-frontend/issues/3932
- Add datadog measurements for NodeJS response time: https://github.com/mozilla/addons-frontend/issues/3755
